gematria/datasets/exegesis_benchmark.cc

```diff
@@ -0,0 +1,325 @@
+// Copyright 2024 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <iostream>
+#include <optional>
+#include <string>
+#include <vector>
+
+#include "X86.h"
+#include "X86InstrInfo.h"
+#include "X86RegisterInfo.h"
+#include "gematria/llvm/disassembler.h"
+#include "gematria/utils/string.h"
+#include "llvm/MC/TargetRegistry.h"
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/Errc.h"
+#include "llvm/Support/Error.h"
+#include "llvm/Support/JSON.h"
+#include "llvm/Support/MemoryBuffer.h"
+#include "llvm/Support/TargetSelect.h"
+#include "llvm/tools/llvm-exegesis/lib/BenchmarkRunner.h"
+#include "llvm/tools/llvm-exegesis/lib/LlvmState.h"
+#include "llvm/tools/llvm-exegesis/lib/ResultAggregator.h"
+#include "llvm/tools/llvm-exegesis/lib/Target.h"
+#include "llvm/tools/llvm-exegesis/lib/TargetSelect.h"
+
+using namespace llvm;
+using namespace llvm::exegesis;
+
+static cl::opt<std::string> AnnotatedBlocksJson(
+    "annotated-blocks-json",
+    cl::desc("Filename of the JSON file containing annotated basic blocks"),
+    cl::init(""));
+
+Expected<BenchmarkCode> parseJSONBlock(
+    const json::Object &BasicBlockJSON, MCInstPrinter &MachinePrinter,
+    const MCDisassembler &MachineDisassembler, const LLVMState &State) {
+  BenchmarkCode BenchCode;
+
+  std::optional<StringRef> HexValue = BasicBlockJSON.getString("Hex");
+  if (!HexValue)
+    return llvm::make_error<StringError>(
+        errc::invalid_argument, "Expected basic block to have hex value");
```

Comment:
This is a general comment, on the whole JSON parsing code, not just this line: we should put in as much information as possible, to allow the user quickly locate the part of the file where the problem is.

In this case, basic block index. For register definitions and memory definitions and mappings, there should be also their index, and so on.

Reply:
I've added block and definition indices pretty much everywhere. Thanks for pushing for this.